openedx-unsupported/edx-analytics-pipeline
# Grades Import Pipeline

**Description**: This PR introduces a series of pipeline tasks that import grades from the LMS for display in Insights.

Specifically, the end result is that two new MySQL tables are created in the analytics DB:
- `grades`: For each course with grades available, this table gives the letter grade, percentage grade, and whether or not the student is passing for every student. This table is intended to be displayed on [the new student profile pages](https://github.com/edx/edx-analytics-dashboard/pull/338) in Insights:
  ![user profile mockup](https://cloud.githubusercontent.com/assets/945577/9422766/4491d766-4858-11e5-9bdf-e91239cc9f6e.png)
  _(don't pay too much attention to the UX above - this is a pipeline PR, not a dashboard/UX PR)_
- `grade_breakdown`: For each course with grades available, this table contains the distribution of letter grades as a function of time. This table is intended to support a new graph in the Insights Dashboard that shows the % of students with a passing grade over time, as a percentage or an absolute number.

Because [grading is slow](https://openedx.atlassian.net/wiki/display/TNL/edX+Grading+Guide+for+Developers), this new pipeline only imports **offline computed grades**. Thus, an instance wishing to use this feature needs to schedule the offline computation of grades to happen [daily] before the analytics pipeline runs. The offline grade computation should be able to cover all courses on small and medium Open edX instances, and could probably be enabled for selected smaller courses on a large instance like edx.org.

**Dependencies**:
- https://github.com/edx/edx-platform/pull/9330 (Fix Offline Grade Calculation)

**Partner information**: not an edX partner - 3rd party-hosted open edX instance

**Test Instructions**:
- In Studio, create a course with a few graded problems. Edit the course grading settings appropriately.
- In the LMS login as various users and complete the problems. Check the grades that each user earns and try to have some users with high grades and others with low grades.
- As the `edxapp` user, compute the offline grades for your course:
  
  ```
  ./manage.py lms compute_grades course-v1:Course+ID+here --settings=devstack
  ```
- Run the pipeline:
  
  ```
  export WHEEL_URL=http://edx-wheelhouse.s3-website-us-east-1.amazonaws.com/Ubuntu/precise
  remote-task --vagrant-path ../ --remote-name devstack --override-config ${PWD}/config/devstack.cfg --repo https://github.com/open-craft/edx-analytics-pipeline.git --branch grades --wheel-url $WHEEL_URL --wait GradesPipelineTask --local-scheduler --n-reduce-tasks 1
  ```
- Check the results using `mysql -u root reports` and running the queries `SELECT * FROM grades;` and `SELECT * FROM grade_breakdown;`.

To reset the pipeline:

```
mysql -u root reports -e 'drop table grade_breakdown; drop table grades; delete from table_updates where target_table IN ("grades", "grade_breakdown");'
hive --service cli -e "drop table grade_breakdown; drop table grades;"
hdfs dfs -rm -r /edx-analytics-pipeline/warehouse/{grade*,offline_grade*}
```

**Reviewers**: @itsjeyd and TBD
